docs/architecture-decision-records/2022-07-module-configuration.md

```diff
@@ -0,0 +1,95 @@
+
+- Status: PROPOSED
+
+## Context
+
+Station modules want to allow Station users to tweak certain aspects of their
+runtime behaviour.
+
+For example, Saturn L2 Node lets users control how much disk space they are
+willing to offer for the cache.
+
+This configuration is module-specific. Each module has different options.
+
+The Station communicates the configuration to the Module via environment
+variables passed to the module backend process.
+
+## Options Considered
+
+### (1) Each module provides config WebUI
+
+The initial idea was to let each Station Module implement its configuration
+screen as part of the module WebUI and persist the configuration in a file. The
+Station would provide the directory where to keep the config file.
+
+We quickly realised this was not optimal. The Saturn module is written in Go and
+expects users to provide all configuration options via environment variables in
+the spirit of [The Twelve-Factor App](https://12factor.net). Implementing a
+configuration layer with GUI requires a lot of work inside Saturn L2 that's not
+adding much value to Saturn:
+
+1. A REST API to read & write the configuration from the WebUI.
+2. Code to persist the configuration in a file. Also, handle possible conflicts
+   between the config file and environment variables.
+3. The config editor GUI.
+4. Support for changing configuration options at runtime with zero downtime.
+
+### (2) Module provides config schema; Station provides configuration WebUI
+
+A different approach is to mimic what VisualStudio Code offers to extensions:
+
+- Modules provide schema describing their configuration, and
+- Station builds a configuration GUI based on the module-provided schema.
+
+This creates a value proposition offered by the Station to future Module
+authors: modules don't have to implement a user-friendly configuration GUI; the
+Station will provide it for free.
+
+There are some caveats, though. Supporting complex configuration formats is a
+lot of work. VS Code has been evolving the configuration editor for years, and
+there are still edge cases where the user has to edit the configuration in the
+JSON file.
+
+Some edge cases to consider:
+
+- A config option is an array of primitive values (an array of strings, an array
+  of numbers)
+- A config option is an enum - there is a set of valid values to pick from.
+- Combine 1+2, and we have an array of enum values.
+- A config option is an object (a key-value map). In some cases, sub-options are
+  well defined - we know what the valid keys are and what value types each key
+  accepts. In other cases, this can be entirely dynamic.
+
+## Decision
+
+We have decided to implement the second option, where the Station provides the
+configuration GUI for Modules.
+
+We will start by supporting:
+
+- string keys mapped directly to environment variables
+- string or integer values
+- optional and required config options
+
+We will implement support for more complex configurations later.
```

Comment:
If there's a robust open source lib that solves our future problems now, it's worth considering starting with that. I'll take some time to look around and report back. 

I like starting simple but like you said wrt vscode config, this could get out of hand in a hurry, and if someone has already solved these problems then we should piggy back off them instead of reinventing the wheel

Reply:
💯

I agree with you, I wish there was a ready-made React component we could use. I did a quick search and did not find any 😞 Maybe you will have more luck 🤞🏻

Reply:
Asked on Twitter: https://twitter.com/bajtos/status/1552194253471744000

Responses so far:

- https://github.com/rjsf-team/react-jsonschema-form
- [nx](https://nx.dev) has that for its plugins. (It's not clear to me where to find the implementation, though.)

Reply:
[react-jsonschema-form](https://github.com/rjsf-team/react-jsonschema-form) looks sweet!

I see one issue, though - they don't support TailwindCSS/daisyUI, which we want to use for visual styling.

Reply:
This is adding complexity that we can address once we need it, compared to mapping key/value pairs directly to html form elements, I think? My only concern was for the schema to be future proof, but I think also there we're good as we're working closely enough with the bleeding-edge modules that this is not an issue yet.